### DISCUSSION TITLE
Analog circuit

### DISCUSSION THREAD

**Discussion by 3hhh** [corrados/edrumulus#98] [General]
As already discussed, I'm pretty sure that the analog circuit is not optimal and likely doesn't work for everyone.
E.g. for me the proposed 22k voltage divider produced a rather bad SNR leading to many other issues.

The differences mostly come from the unknown Piezo output voltage range, which may be +- 0,1-10V.
The effective voltage range depends on:
- piezo hardware (size, thickness)
- trigger cushion foam (i.e. the maximum velocity at which a user can possibly hit the piezo)

In my particular case the voltage range of the piezo seems to be around +-1V and rarely more. This can be measured rather easily by plugging the piezo to the line in which runs at +-1V range. If there's no clipping, then you're apparently in that range.

I played a bit with the apparent industry-standard software [LTSpice](https://www.analog.com/en/design-center/design-tools-and-calculators/ltspice-simulator.html) to find a set of parameters that may work for me and interestingly found quite a different one from the proposed parameters.

Here's a comparison of the parameters proposed by edrumulus and the seemingly better ones for my case (+-1V, 2kHz spikes):
![2](https://github.com/corrados/edrumulus/assets/13033030/01235adb-8f51-4c39-b8a1-ecacdcf75947)
![1](https://github.com/corrados/edrumulus/assets/13033030/85c4276f-148d-4391-af1e-50ecbe3da3f3)
[simulation (needs to be renamed to edrumulus.asc)](https://github.com/corrados/edrumulus/files/12576561/edrumulus.txt)
(The C3=1p case is roughly identical to what edrumulus does = no capacitor.)

One can see that the voltage range in the edrumulus example is ~0,3V, whereas it could be ~2V as in the second example. Of course that also affects the SNR.

Therefore imho it would be more honest to tell people that they need to play with the parameters to find a circuit that may work for them.

Disclaimer: I still need to validate the simulation in practice.

**Comment by corrados** [comment]
> If there's no clipping, then you're apparently in that range.

Some clipping is allowed. I have implemented a clipping compensation in software.

> Therefore imho it would be more honest to tell people that they need to play with the parameters to find a circuit that may work for them.

This is the current disclaimer on the main page:

**DISCLAIMER All these circuit diagrams are used for my first prototypes and work fine for me. Use them at your own risk.**

If you want to see a different text then please make a proposal.

**Comment by corrados** [comment]
BTW, I may convert this Issue in a Discussion thread later on because as soon as I close it, most people will not find the investigation results you have given in your initial post. I assume that new users will only check for help in the Discussion forums.

**Comment by 3hhh** [comment]
> This is the current disclaimer on the main page:
> 
> **DISCLAIMER All these circuit diagrams are used for my first prototypes and work fine for me. Use them at your own risk.**
> 
> If you want to see a different text then please make a proposal.

I guess "In particular the circuits may not work for currently unsupported pieces of hardware." might be fair enough for a disclaimer.

However additionally some documentation on how to support new pieces of hardware would've been clarifying and helpful.

Currently the steps appear to be:

1. Start with prototype2 & optimize the analog circuit (atm it is unclear to me how to do that apart from random testing; I might know more once I can verify my simulation in practice).
2. Create samples for the edrumulus Octave model according to the existing doc. Probably getting them from the edrumulus hardware itself is more realisitic than from whatever sound card.
3. Analyze the samples in the Octave model to identify the edrumulus parameters. We discussed that in the forum.
4. Set the identified parameters in the edrumulus C++ code and test them in practice. Modify the available real-time parameters for hopefully better results. Possibly go back to previous steps.

**Comment by corrados** [comment]
> I guess "In particular the circuits may not work for currently unsupported pieces of hardware." might be fair enough for a disclaimer.

done

**Comment by corrados** [comment]
I have added your step-by-step guide to https://github.com/corrados/edrumulus/blob/main/doc/manual.md. Feel free to correct/improve the text.

**Comment by 3hhh** [comment]
Nice, thanks!

Btw I tested it in practice and unfortunately I couldn't record a signal that looked significantly better than previous ones (maybe there's some generic issue with the way I'm recording, idk). However I could lower the edrumulus threshold from an unusable 20 to around 10. Therefore I guess I'm going to stick with Rs=1k, Rp=100k, Cp=0 for the MPS 750x Tom for now. Cp dampens the gain as expected, but it doesn't look like I need it for the MPS 750x Tom.

**Comment by 3hhh** [comment]
> Therefore I guess I'm going to stick with Rs=1k, Rp=100k, Cp=0 for the MPS 750x Tom for now.

Since all the other line-in samples have little clipping, I'd guess that these parameters apply to all previously supported pads, too.

Therefore it would be great if someone verfied my claims in practice with an oscilloscope.
If the claims are correct, it would make sense to adjust the general recommendations and circuits.

**Comment by 3hhh** [comment]
By now I believe that Rs=10k, Rp=100k, Cp=0 is more safe to use with e.g. the Teensy. It is also absolutely necessary to use BAT85 or similar diodes with low forward voltage to limit the Teensy input to its ADC operating range of -0,3V - 3,6V max. The 1N4148 is e.g. **not** suitable (contrary to the circuit above) as it has ~0,7V forward voltage at 1mA leading to ~4V at the Teensy, whereas the BAT85 has 0,3V forward voltage at that current (--> 3,6V at the Teensy).

Rs=1k would mean 10mA at 10V piezo input voltage resulting in 0,4V forward voltage for the BAT85 and thus 3,7V at the Teensy, which is out of its operating range. Therefore Rs=10k is a bit more safe whilst only sacrificing 10% signal input.

(@corrados I decided to continue our circuit discussion here.)

**Comment by corrados** [comment]
In the end, it's your choice how you design your prototype :-)

Here is a link to a discussion with jstma who really knows how to design hardware (in contrast to me since I am a software developer): https://github.com/corrados/edrumulus/discussions/58#discussioncomment-3449568

**Reply by 3hhh** [reply]
> Here is a link to a discussion with jstma who really knows how to design hardware (in contrast to me since I am a software developer): [#58 (reply in thread)](https://github.com/corrados/edrumulus/discussions/58#discussioncomment-3449568)

An interesting thread.

Regarding the diodes @jstma confirms that they are rather pointless at Rs=100k, Rp=22k (after all you need roughly 3,3V *122/22 = 18,3V to trigger them, which is rather unlikely for a Piezo input...). It's different for my circuit with Rs=10k, Rp=100k --> 3,6V though as long as one fears 10V piezos out there.

I also had a look at the frequency response and didn't notice any disadvantage in choosing Rs=10k over Rs=100k:
![freq_resp](https://github.com/corrados/edrumulus/assets/13033030/f76e5b1a-e4d0-477d-9dfa-3dd22f9c2e11)
(red is 100k)

What I see is a phase shift at low frequencies, which may distort microphones or maybe guitar pickups. However it should be irrelevant to e-drums.

But maybe I did something wrong - I'm a newbie to LTSpice after all...

Btw regarding your hihat control circuit it might make sense to put a note somewhere in the manual that you're expecting the hihat controller to be a poti. According to my calculations you're apparently expecting `poti ~ 2*R38` (in the [prototype 2 circuit](https://github.com/corrados/edrumulus/blob/main/hardware/prototype2/edrumulus.pdf)), i.e. it may make sense for users to measure their poti with a multimeter.

Also please don't get me wrong:
I like what you do. But since I'm likely going to stick with prototype 2 for maximum flexibility regarding my drum set, I'm trying to improve it and make it possible to adapt arbitrary hardware to it.

**Comment by 3hhh** [comment]
The new Teensy works as expected - I tested its ADCs by reading from a constant 1,65V voltage divider.

However now I unfortunately hit into a new issue:
The drum pads draw current. Thus if I run with Rs+Rp=110k (or in the prototype2 circuit 122k) between 3V3 and the pad there won't be enough current and the 1,65V voltage divider voltage will go down to 0,2V or so...
You appear to ignore that issue in your code by computing a DC offset at edrumulus startup (which _should_ be 1,65V, but in reality probably mostly isn't).
Or don't Roland pads draw current?
Possibly it's also the ADC when it's reading...

I'll attempt to fix it by using Rs = Rp = 10k and allow 3V3/20k = 0,165mA (which isn't much), but of course that'll drop 50% signal over Rs, which isn't optimal either...
So either I may lose the negative signal range or lose 50% over Rs...
To me it looks like this passive circuit is at its limits.

**Reply by corrados** [reply]
> Or don't Roland pads draw current?

They don't. If you look at [my recordings](https://github.com/corrados/edrumulus/blob/main/doc/ChangeLog.md#2021-04-04-adc-spikes-of-the-esp32), you can see that the values are at approx. 1800 which are close to the middle point of the ADC which is at 2048.

Your Millenium pads behave really strange. In Roland pads, usually the piezo sensor is directly connected to the audio jack. I thought this would also be the case for your pads.

**Reply by 3hhh** [reply]
The dynamics with edrumulus on my new Teensy (MPS 750x with Rs=Rp=10k) are quite impressive! Many thanks for your project again @corrados! :-)

I can still see the BAT85 diodes firing (= Teensy blinking due to short circuit) from time to time when I hit hard, but it shouldn't be an issue. I don't need the dynamics between "very hard" and "very very hard" anyway...
So it looks like at least some of the MPS piezo outputs reach +-2V or so.

Nonetheless I still have a couple of issues, most notably a lot of cross talk - especially with the snare.
So how does the cross-talk parameter work?

P.S.: I'll open separate issues for the other minor ones.

**Reply by corrados** [reply]
Have you tried my values of Rs and Rp? Does this also work ok?

What is your threshold setting? I could go pretty low with the Teensy. On the ESP32 developer board I had to use higher values.

Cross talk works in a way that you increase this parameter to block hits from other pads. The higher the value, the higher the peaks can be which might be suppressed. If that value is too high, regular hits on that pad might be suppressed as well which is not desired.

**Reply by 3hhh** [reply]
> Have you tried my values of Rs and Rp? Does this also work ok?

No, I didn't see the point of trying as my Rs+Rp=110k to 3V3 had produced the voltage divider voltage drop. So I doubt it would be better with Rs+Rp=122k.
edrumulus might work somehow with 0,2V, but it would be a bad idea nonetheless.

> What is your threshold setting? I could go pretty low with the Teensy. On the ESP32 developer board I had to use higher values.

For snare & toms I used a threshold of 2, for the cymbals I went higher to dampen the signal, but there's no noise at lower thresholds (maybe I should rather lower the sensitivity...).

> Cross talk works in a way that you increase this parameter to block hits from other pads. The higher the value, the higher the peaks can be which might be suppressed. If that value is too high, regular hits on that pad might be suppressed as well which is not desired.

So if I play a pad and get cross-talk, I need to increase that pad's cross-talk setting?
Ok, I'll try that.

**Reply by 3hhh** [reply]
Hm even cross talk = 31 doesn't change much. I'll have to test this more...

**Reply by 3hhh** [reply]
After running with Rs=Rp=10k for a while I noticed that the hot spot effect when hitting directly on the Tom piezo causes edrumulus to go mute. It doesn't happen on all toms. I guess it's because some piezos have negative polarity and the diode shorting to 0V causes this.

Ironically I also noticed that I kind of like it as I have the piezos at the top of the pads and hitting the snare mesh there causes me to get a rimshot as the hit piezo goes to 0V, but the rim piezo still rings - this is way better than a rim shot on the actual rim, which causes this annoying plastic sound... :-)

**Reply by corrados** [reply]
> After running with Rs=Rp=10k for a while I noticed that the hot spot effect when hitting directly on the Tom piezo causes edrumulus to go mute. It doesn't happen on all toms. I guess it's because some piezos have negative polarity and the diode shorting to 0V causes this.

It would be interesting to record such an incident (ADC signal) and analyze the effect in Matlab. Then you could see for sure what's actually happening.
